packages/jsii/lib/jsii-diagnostic.ts

```diff
@@ -816,3 +816,42 @@ export class JsiiDiagnostic implements ts.Diagnostic {
 export type DiagnosticMessageFormatter = (
   ...args: any[]
 ) => JsiiDiagnostic['messageText'];
+
+export function configureCategories(records: { [code: string]: string }) {
+  for (const [code, level] of Object.entries(records)) {
+    const diagCode = Code.lookup(diagnosticCode(code));
+    const category = diagnosticCategory(level);
+    if (!diagCode) {
+      throw new Error(`Unrecognized diagnostic code '${code}'`);
+    }
+    diagCode.category = category;
+  }
+}
+
+function diagnosticCategory(str: string): ts.DiagnosticCategory {
+  switch (str.trim().toLowerCase()) {
+    case 'error':
+      return ts.DiagnosticCategory.Error;
+    case 'warning':
+      return ts.DiagnosticCategory.Warning;
+    case 'suggestion':
+      return ts.DiagnosticCategory.Suggestion;
+    case 'message':
+      return ts.DiagnosticCategory.Message;
+    default:
+      throw new Error(`Unrecognized diagnostic category '${str}'`);
+  }
+}
+
+function diagnosticCode(str: string): string | number {
+  if (str.toLowerCase().startsWith('jsii')) {
+    const num = +str.slice(4);
+    if (Number.isNaN(num)) {
+      throw new Error(
+        `Invalid diagnostic code ${str}. A number must follow code that starts with 'JSII'`,
+      );
+    }
+    return num;
+  }
```

Comment:
Ooof! This looks a little too brittle IMHO.

```suggestion
  const m = /^JSII(\d+)$/i.exec(str);
  if (m) {
    return parseInt(m[1], 10);
  }
```

In case where `JSII` is not followed by digits (where it might end up NAN, but could also have been a float per your test), I'd just use it up as a string, and let that fail.